src/core/compiler.rkt

```diff
@@ -47,18 +45,25 @@
     [(list op a b c) (op (vector-ref regs a) (vector-ref regs b) (vector-ref regs c))]
     [(list op args ...) (apply op (map (curry vector-ref regs) args))]))
 
-; This functions needs to preserve vars
-(define (batch-for-compiler batch brfs vars)
-  (define-values (batch* brfs*)
-    (batch-apply batch
-                 brfs
-                 (lambda (node)
-                   (match node
-                     [(approx spec impl) impl]
-                     [node node]))))
-  (define-values (batch** _) (progs->batch vars))
-  (define brfs** (map (batch-copy-only! batch** batch*) brfs*))
-  (values batch** brfs**))
+;; This function:
+;;   1) copies only nodes associated with provided brfs - so, gets rid of useless nodes
+;;   2) rewrites these nodes as fl-instructions
+(define (batch-for-compiler batch brfs vars args)
+  (define out (batch-empty))
+  (define f
+    (batch-recurse
+     batch
+     (λ (brf recurse)
+       (match (deref brf)
+         [(approx _ impl) (recurse impl)] ;; do not push, it is already a batchref
+         [(? symbol? n)
+          (define idx (index-of vars n))
+          (batch-push! out (list (λ () (vector-ref args idx))))]
```

Comment:
So I'm a little confused, but I think instead of making the `args` the first few cells of the `vregs`, you're putting them in a separate array and adding an instruction to copy them? Why?